File: src/main/java/x/withlithum/neoware/data/game/ItemDefinition.java
package x.withlithum.neoware.data.game;

import lombok.extern.slf4j.Slf4j;
import net.kyori.adventure.key.Key;
import net.kyori.adventure.nbt.CompoundBinaryTag;
import net.kyori.adventure.nbt.StringBinaryTag;
import net.kyori.adventure.text.Component;
import net.minestom.server.MinecraftServer;
import net.minestom.server.codec.Codec;
import net.minestom.server.codec.StructCodec;
import net.minestom.server.component.DataComponent;
import net.minestom.server.component.DataComponents;
import net.minestom.server.item.ItemStack;
import net.minestom.server.item.Material;
import net.minestom.server.item.component.AttributeList;
import net.minestom.server.item.component.CustomData;
import net.minestom.server.item.component.EnchantmentList;
import net.minestom.server.item.component.TooltipDisplay;
import net.minestom.server.item.enchant.Enchantment;
import net.minestom.server.registry.RegistryKey;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;
import x.withlithum.neoware.data.encode.MiniMessageCodec;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

@Slf4j
@NullMarked
public record ItemDefinition(Material material,
                             @Nullable Component name,
                             @Nullable List<Component> lore,
                             @Nullable Map<Key, Integer> enchantments,
                             @Nullable AttributeList attributeModifiers,
                             @Nullable Set<DataComponent<?>> hideTooltips) {
    public static final Codec<ItemDefinition> CODEC = StructCodec.struct(
        "material", Material.CODEC, ItemDefinition::material,
        "name", MiniMessageCodec.INSTANCE.optional(), ItemDefinition::name,
        "lore", MiniMessageCodec.INSTANCE.list(256).optional(), ItemDefinition::lore,
        "enchantments", Codec.KEY.mapValue(Codec.INT), ItemDefinition::enchantments,
        "attribute_modifiers", AttributeList.CODEC.optional(), ItemDefinition::attributeModifiers,
        "hide_tooltips", DataComponent.CODEC.set().optional(), ItemDefinition::hideTooltips,
        ItemDefinition::new
    );

    public ItemStack createItem(Key id) {
        var builder = ItemStack.builder(material);
        builder.set(DataComponents.CUSTOM_DATA, new CustomData(CompoundBinaryTag.builder()
            .put("base", StringBinaryTag.stringBinaryTag(id.asString()))
            .build()));
        if (name != null) {
            builder.set(DataComponents.ITEM_NAME, name);
        }
        if (lore != null) {
            builder.lore(lore);
        }
        if (enchantments != null) {
            builder.set(DataComponents.ENCHANTMENTS, createEnchantmentList(id, enchantments));
        }
        if (attributeModifiers != null) {
            builder.set(DataComponents.ATTRIBUTE_MODIFIERS, attributeModifiers);
        }
        if (hideTooltips != null) {
            builder.set(DataComponents.TOOLTIP_DISPLAY, new TooltipDisplay(false,
                hideTooltips));
        }
        return builder.build();
    }

    private static EnchantmentList createEnchantmentList(Key defId,
                                                         Map<Key, Integer> enchantments) {
        var el = new HashMap<RegistryKey<Enchantment>, Integer>(enchantments.size());
        var registry = MinecraftServer.getEnchantmentRegistry();

        for (var enchantId : enchantments.keySet()) {
            var regKey = registry.getKey(enchantments.get(enchantId));
            if (regKey == null) {
                log.warn("Definition '{}' has non-existent enchantment '{}' declared",
                    defId,
                    enchantId);
                continue;
            }

            el.put(regKey, enchantments.get(enchantId));
        }

        return new EnchantmentList(el);
    }
}

File: src/main/kotlin/x/withlithum/neoware/data/game/ItemDefinition.kt
/*
 * SPDX-FileCopyrightText: 2025-2026 WithLithum & contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

package x.withlithum.neoware.data.game

import kotlinx.serialization.Serializable
import net.kyori.adventure.key.Key
import net.kyori.adventure.nbt.CompoundBinaryTag
import net.kyori.adventure.nbt.StringBinaryTag
import net.kyori.adventure.text.Component
import net.minestom.server.MinecraftServer
import net.minestom.server.component.DataComponent
import net.minestom.server.component.DataComponents
import net.minestom.server.entity.attribute.AttributeModifier
import net.minestom.server.item.ItemStack
import net.minestom.server.item.Material
import net.minestom.server.item.component.AttributeList
import net.minestom.server.item.component.CustomData
import net.minestom.server.item.component.EnchantmentList
import net.minestom.server.item.component.TooltipDisplay
import net.minestom.server.item.enchant.Enchantment
import net.minestom.server.registry.RegistryKey
import x.withlithum.neoware.data.serial.DataComponentTypeSerializer
import x.withlithum.neoware.data.serial.KeySerializer
import x.withlithum.neoware.data.serial.MaterialSerializer
import x.withlithum.neoware.data.serial.MiniMessageSerializer
import x.withlithum.neoware.util.ifNotNull

@JvmRecord
@Serializable
data class ItemDefinition(
    @Serializable(with = MaterialSerializer::class)
    val material: Material,
    @Serializable(with = MiniMessageSerializer::class)
    val name: Component? = null,
    val lore: List<@Serializable(MiniMessageSerializer::class) Component>? = null,
    val enchantments: Map<@Serializable(KeySerializer::class) Key, Int>? = null,
    val attributes: List<ItemAttributeRecord>? = null,
    val hideTooltips: Set<@Serializable(DataComponentTypeSerializer::class) DataComponent<*>>? = null
) {
    companion object {
        private fun createEnchantmentList(map: Map<Key, Int>): EnchantmentList {
            val el = HashMap<RegistryKey<Enchantment>, Int>(map.size)
            val registry = MinecraftServer.getEnchantmentRegistry()

            for (entry in map.entries) {
                val regKey = registry.getKey(entry.key) ?: throw IllegalStateException("Unknown enchantment type '${entry.key}'")
                el[regKey] = entry.value
            }

            return EnchantmentList(el)
        }

        private fun createAttributeList(records: List<ItemAttributeRecord>): AttributeList {
            val al = ArrayList<AttributeList.Modifier>(records.size)
            for (entry in records) {
                val display = if (entry.hide) {
                    AttributeList.Display.Hidden.INSTANCE
                } else {
                    AttributeList.Display.Default.INSTANCE
                }

                val item = AttributeList.Modifier(entry.attribute,
                    AttributeModifier(entry.id, entry.amount, entry.operation),
                    entry.slot,
                    display)
                al.add(item)
            }

            return AttributeList(al)
        }

    }

    fun createItem(baseId: Key): ItemStack {
        val builder = ItemStack.builder(material)
        builder.set(DataComponents.CUSTOM_DATA, CustomData(CompoundBinaryTag.builder()
            .put("base", StringBinaryTag.stringBinaryTag(baseId.asString()))
            .build()))
        name.ifNotNull { builder.set(DataComponents.ITEM_NAME, it) }
        lore.ifNotNull { builder.set(DataComponents.LORE, it) }
        enchantments.ifNotNull { builder.set(DataComponents.ENCHANTMENTS, createEnchantmentList(it)) }
        attributes.ifNotNull { builder.set(DataComponents.ATTRIBUTE_MODIFIERS, createAttributeList(it)) }
        hideTooltips.ifNotNull { builder.set(DataComponents.TOOLTIP_DISPLAY, TooltipDisplay(false,
            it)) }
        return builder.build()
    }

}
